File: cmd/pull.go
package cmd

import (
	"fmt"
	"os"

	"github.com/aura-studio/dynamic-node-cli/config"
	"github.com/aura-studio/dynamic-node-cli/pull"
	"github.com/spf13/cobra"
)

// pullCmd represents the pull command.
var pullCmd = &cobra.Command{
	Use:   "pull",
	Short: "Pull artifacts from S3 warehouse to local",
	Long:  "Reads dynamic-node-cli.yaml and the given --procedure, then pulls .zip artifacts from remote(s) to local warehouse.",
	Run: func(cmd *cobra.Command, args []string) {
		cfgPath := resolveConfigPath(cmd)

		proc, err := cmd.Flags().GetString("procedure")
		if err != nil {
			fmt.Println("error:", err)
			os.Exit(1)
		}
		concurrency, err := cmd.Flags().GetInt("concurrency")
		if err != nil {
			fmt.Println("error:", err)
			os.Exit(1)
		}
		force, err := cmd.Flags().GetBool("force")
		if err != nil {
			fmt.Println("error:", err)
			os.Exit(1)
		}
		remote, err := cmd.Flags().GetString("remote")
		if err != nil {
			fmt.Println("error:", err)
			os.Exit(1)
		}

		c := config.Parse(cfgPath)
		config.Validate(c)

		opt := pull.Options{Concurrency: concurrency, Force: force, Remote: remote}

		if proc == "" {
			fmt.Println("No procedure specified, pulling all procedures...")
			procedures := config.GetAllProcedures(c)
			for _, procName := range procedures {
				fmt.Printf("\nPulling procedure: %s\n", procName)
				procObj := config.CreateProcedure(c, procName)
				pull.PullForProcedure(procObj, opt)
			}
			fmt.Println("\nAll procedures pulled successfully.")
			return
		}

		procObj := config.CreateProcedure(c, proc)
		pull.PullForProcedure(procObj, opt)
	},
}

func init() {
	rootCmd.AddCommand(pullCmd)
	pullCmd.Flags().StringP("config", "c", "", "path to dynamic-node-cli.yaml (default: ./dynamic-node-cli.yaml or ./dynamic-node-cli.yml)")
	pullCmd.Flags().StringP("procedure", "p", "", "procedure name to pull (optional, pulls all if not specified)")
	pullCmd.Flags().IntP("concurrency", "j", 8, "max concurrent downloads per remote")
	pullCmd.Flags().BoolP("force", "f", false, "overwrite existing local files")
	pullCmd.Flags().String("remote", "", "override warehouse.remote (pull from this remote only)")
}

File: src/cmd/pull.js
import { Command } from "commander";
import { parseConfig, validateConfig } from "../config/config.js";
import { createProcedure, getAllProcedures } from "../config/procedure.js";
import { pullForProcedure } from "../pull/pull.js";
import { resolveConfigPath } from "./config-path.js";

export function createPullCommand() {
  return new Command("pull")
    .description("Pull artifacts from S3 warehouse to local")
    .option(
      "-c, --config <path>",
      "path to dynamic-node-cli.yaml (default: ./dynamic-node-cli.yaml or ./dynamic-node-cli.yml)",
      "",
    )
    .option(
      "-p, --procedure <name>",
      "procedure name to pull (optional, pulls all if not specified)",
      "",
    )
    .option(
      "-j, --concurrency <number>",
      "max concurrent downloads per remote",
      (value) => Number.parseInt(value, 10),
      8,
    )
    .option("-f, --force", "overwrite existing local files", false)
    .option("--remote <remote>", "override warehouse.remote (pull from this remote only)", "")
    .action(async (options) => {
      const cfgPath = resolveConfigPath(options);
      const config = parseConfig(cfgPath);
      validateConfig(config);

      const opt = {
        concurrency: options.concurrency,
        force: options.force,
        remote: options.remote,
      };

      if (!options.procedure) {
        console.log("No procedure specified, pulling all procedures...");
        const procedures = getAllProcedures(config);
        for (const procName of procedures) {
          console.log(`\nPulling procedure: ${procName}`);
          const procObj = createProcedure(config, procName);
          await pullForProcedure(procObj, { ...opt });
        }
        console.log("\nAll procedures pulled successfully.");
        return;
      }

      const procObj = createProcedure(config, options.procedure);
      await pullForProcedure(procObj, opt);
    });
}
